lib/contract/invoke_host_function.ex

```diff
@@ -0,0 +1,162 @@
+defmodule Soroban.Contract.InvokeHostFunction do
+  @moduledoc """
+  `InvokeHostFunction` implementation to invoke authorized and not authorized contract functions.
+  """
+
+  alias Soroban.RPC
+  alias Soroban.RPC.{SendTransactionResponse, SimulateTransactionResponse}
+  alias Stellar.{Horizon.Accounts, TxBuild}
+
+  alias Stellar.TxBuild.{
+    Account,
+    ContractAuth,
+    HostFunction,
+    InvokeHostFunction,
+    SCVal,
+    SequenceNumber,
+    Signature
+  }
+
+  @type account :: Account.t()
+  @type function_args :: list(struct())
+  @type auth :: String.t() | nil
+  @type auth_account :: String.t() | nil
+  @type auth_accounts :: list(binary())
+  @type invoke_host_function :: InvokeHostFunction.t()
+  @type function_name :: String.t()
+  @type contract_id :: binary()
+  @type source_secret_key :: binary()
+  @type simulate_response :: {:ok, SimulateTransactionResponse.t()}
+  @type send_response :: {:ok, SendTransactionResponse.t()}
+  @type signature :: Signature.t()
+  @type sequence_number :: SequenceNumber.t()
+  @type sc_val_list :: list(SCVal.t())
+
+  @spec invoke(
+          contract_id :: contract_id(),
+          source_secret_key :: source_secret_key(),
+          function_name :: function_name(),
+          function_args :: function_args(),
+          auth_accounts :: auth_accounts()
+        ) :: send_response()
+  def invoke(
+        contract_id,
+        source_secret_key,
+        function_name,
+        function_args,
+        auth_accounts \\ []
+      ) do
+    function_args = covert_to_sc_val(function_args)
+    keypair = {public_key, _secret} = Stellar.KeyPair.from_secret_seed(source_secret_key)
+    signature = Signature.new(keypair)
+    source_account = Account.new(public_key)
+    {:ok, seq_num} = Accounts.fetch_next_sequence_number(public_key)
+    sequence_number = SequenceNumber.new(seq_num)
```

Comment:
@Odraxs the use of a `with` clause will make your function response and error handling predictable.